src/Agent.Worker/TaskManager.cs

```diff
@@ -308,6 +315,44 @@ private async Task DownloadAsync(IExecutionContext executionContext, Pipelines.T
             }
         }
 
+        private void CheckForTaskDeprecation(IExecutionContext executionContext, Pipelines.TaskStepDefinitionReference task)
+        {
+            string taskJsonPath = Path.Combine(GetDirectory(task), "task.json");
+            string taskJsonText = File.ReadAllText(taskJsonPath);
+            JObject taskJson = JObject.Parse(taskJsonText);
+            var deprecated = taskJson["deprecated"];
+
+            if (deprecated != null && deprecated.Value<bool>())
+            {
+                string friendlyName = taskJson["friendlyName"].Value<string>();
+                int majorVersion = new Version(task.Version).Major;
+                string deprecationMessage = $"Task '{friendlyName}' version {majorVersion} ({task.Name}@{majorVersion}) is deprecated";
+                var removalDate = taskJson["removalDate"];
+
+                if (removalDate != null)
+                {
+                    string removalDateString = removalDate.Value<DateTime>().ToString("MMMM d, yyyy");
+                    deprecationMessage += $" and will be removed. From {removalDateString} onwards it may no longer be available";
```

Comment:
Should these strings should be localized?

https://github.com/microsoft/azure-pipelines-agent/pull/4454